OpenRA.Mods.Common/Traits/Render/Hovers.cs

```diff
@@ -19,35 +19,71 @@
 namespace OpenRA.Mods.Common.Traits.Render
 {
 	[Desc("Changes the visual Z position periodically.")]
-	public class HoversInfo : ConditionalTraitInfo, Requires<IMoveInfo>
+	public class HoversInfo : PausableConditionalTraitInfo, Requires<IMoveInfo>
 	{
 		[Desc("Amount of Z axis changes in world units.")]
-		public readonly int OffsetModifier = -43;
+		public readonly int OffsetModifier = 64;
 
 		public readonly int MinHoveringAltitude = 0;
 
+		[Desc("Amount of ticks it takes to reach OffsetModifier.")]
+		public readonly int Ticks = 25;
+
+		[Desc("Amount of ticks it takes to fall to the ground from the highest point when disabled.")]
+		public readonly int FallTicks = 12;
+
+		[Desc("Amount of ticks it takes to rise from the ground to InitialHeight.")]
+		public readonly int RiseTicks = 17;
+
+		public readonly int InitialHeight = 384;
+
 		public override object Create(ActorInitializer init) { return new Hovers(this, init.Self); }
 	}
 
-	public class Hovers : ConditionalTrait<HoversInfo>, IRenderModifier
+	public class Hovers : PausableConditionalTrait<HoversInfo>, IRenderModifier, ITick
 	{
 		readonly HoversInfo info;
+		readonly float stepPercentage;
+		readonly float fallTickHeight;
+		int ticks = 0;
+		WVec worldVisualOffset = WVec.Zero;
 
 		public Hovers(HoversInfo info, Actor self)
 			: base(info)
 		{
 			this.info = info;
+			this.stepPercentage = 90f / info.Ticks * (float)Math.PI / 180f;
+			this.fallTickHeight = (info.InitialHeight + info.OffsetModifier) / info.FallTicks;
 		}
 
-		IEnumerable<IRenderable> IRenderModifier.ModifyRender(Actor self, WorldRenderer wr, IEnumerable<IRenderable> r)
+		void ITick.Tick(Actor self)
 		{
-			if (self.World.Paused || IsTraitDisabled)
-				return r;
+			if (IsTraitDisabled || IsTraitPaused)
+			{
+				if (worldVisualOffset.Z < 0)
+					return;
 
-			var visualOffset = self.World.Map.DistanceAboveTerrain(self.CenterPosition).Length >= info.MinHoveringAltitude
-				? (int)Math.Abs((self.ActorID + Game.LocalTick) / 5 % 4 - 1) - 1
+				var fallTicks = worldVisualOffset.Z / fallTickHeight - 1;
+				worldVisualOffset = new WVec(0, 0, (int)(fallTickHeight * fallTicks));
+			}
+			else
+				ticks++;
+		}
+
+		IEnumerable<IRenderable> IRenderModifier.ModifyRender(Actor self, WorldRenderer wr, IEnumerable<IRenderable> r)
+		{
+			if (!IsTraitDisabled && !IsTraitPaused)
+			{
+				var visualOffset = self.World.Map.DistanceAboveTerrain(self.CenterPosition).Length >= info.MinHoveringAltitude
+				? Math.Sin(ticks % (info.Ticks * 4) * stepPercentage)
```

Comment:
Please use `WAngle` and `WAngle.Sin()` instead of floating point and `System.Math`.